pkg/controller/baremetalhost/host_state_machine.go

```diff
@@ -135,21 +130,43 @@ func (hsm *hostStateMachine) checkInitiateDelete() bool {
 	return true
 }
 
-func (hsm *hostStateMachine) shouldInitiateRegister(info *reconcileInfo) bool {
-	changeState := false
-	if hsm.Host.DeletionTimestamp.IsZero() {
-		switch hsm.NextState {
-		default:
-			changeState = !hsm.Host.Status.GoodCredentials.Match(*info.bmcCredsSecret)
-		case metal3v1alpha1.StateNone, metal3v1alpha1.StateUnmanaged:
-		case metal3v1alpha1.StateRegistering, metal3v1alpha1.StateRegistrationError:
-		case metal3v1alpha1.StateDeleting:
+func (hsm *hostStateMachine) ensureRegistered(info *reconcileInfo) (result actionResult) {
+	switch hsm.NextState {
+	case metal3v1alpha1.StateNone, metal3v1alpha1.StateUnmanaged:
+		// We haven't yet reached the Registration state, so don't attempt
+		// to register the Host.
+		return
+	}
+	if !hsm.Host.DeletionTimestamp.IsZero() {
+		// BUG(zaneb) We currently don't attempt to re-register the Host
+		// if we find it missing once a delete has been requested (in
+		// part because we are also willing to pass empty credentials
+		// after this point), but this means that we may not be able to
+		// deprovision a Host if the Ironic database pod is rescheduled
+		// during deprovisioning, or if the credentials need to be
+		// modified.
+		return
+	}
+
+	if hsm.Host.Status.GoodCredentials.Match(*info.bmcCredsSecret) {
```

Comment:
Would this mean that a host that was successfully registered once is not re-registered if the database is wiped after a pod restart? Maybe that's handled the next time the provisioner is used to try to do something with the host?

Elsewhere we say that after a host passes through the registration state it never goes back. I wonder if we need this check at all any more?

Reply:
> Would this mean that a host that was successfully registered once is not re-registered if the database is wiped after a pod restart?

It means that we won't call actionRegistering below.
In the previous code, it meant we wouldn't flip back to the Registering state (which calls actionRegistering).
So the effect is the same, which is **not** to say that there isn't a bug. I suspect there are a lot of states where we don't re-register the host if the DB is dropped. (The ones that call Adopt() are the only ones where we _do_ AFAIK.)

> Elsewhere we say that after a host passes through the registration state it never goes back.

This is more of an accounting thing. We do re-register below if the creds change, by doing the exact stuff that the Registering state does, but we don't change the provisioning state while we do it any more.